secrets/secrets-vault/src/lib.rs

```diff
@@ -0,0 +1,497 @@
+    jwks_address: SocketAddrV4,
+    subject_mapper: SubjectMapper,
+    vault_config: VaultConfig,
+}
+
+pub struct VaultConfig {
+    pub address: String,
+    pub auth_mount: String,
+    pub audience: String,
+}
+
+impl VaultSecretsBackend {
+    pub fn new(
+        nats_client: async_nats::Client,
+        nkey: KeyPair,
+        xkey: XKey,
+        jwks_address: SocketAddrV4,
+        subject_mapper: SubjectMapper,
+        vault_config: VaultConfig,
+    ) -> Self {
+        Self {
+            nats_client,
+            nkey,
+            xkey,
+            jwks_address,
+            subject_mapper,
+            vault_config,
+        }
+    }
+
+    pub async fn serve(&self) -> anyhow::Result<()> {
+        let pk = KeyPair::from_public_key(&self.nkey.public_key())?;
+        tokio::spawn({
+            let listen_address = self.jwks_address.to_owned();
+            async move {
+                VaultSecretsJwksServer::new(vec![pk], listen_address)
+                    .unwrap()
+                    .serve()
+                    .await
+                    .unwrap()
+            }
+        });
+        self.start_nats_subscriber().await?;
+        Ok(())
+    }
+
+    async fn start_nats_subscriber(&self) -> Result<()> {
+        debug!(
+            "Subscribing to messages addressed to {} under queue group {}",
+            self.subject_mapper.secrets_wildcard_subject(),
+            self.subject_mapper.queue_group_name(),
+        );
+
+        let subject_prefix = self.subject_mapper.secrets_subject();
+        let mut subscriber = self
+            .nats_client
+            .queue_subscribe(
+                self.subject_mapper.secrets_wildcard_subject(),
+                self.subject_mapper.queue_group_name(),
+            )
+            .await?;
+
+        while let Some(message) = subscriber.next().await {
+            // We check to see if there's a reply inbox, otherwise just ignore the message.
+            let Some(reply_to) = message.reply.clone() else {
+                continue;
+            };
+
+            match parse_op_from_subject(&message.subject, &subject_prefix) {
+                Ok(Operation::Get) => {
+                    if let Err(err) = self.handle_get_request(reply_to.clone(), message).await {
+                        self.handle_get_request_error(reply_to, err).await;
+                    }
+                }
+                Ok(Operation::ServerXkey) => {
+                    self.handle_server_xkey_request(reply_to).await;
+                }
+                Err(err) => {
+                    self.handle_unsupported_request(reply_to, err).await;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_unsupported_request(&self, reply_to: Subject, error: anyhow::Error) {
+        // TODO: handle the potential publish error
+        let _ = self
+            .nats_client
+            .publish(reply_to, error.to_string().into())
+            .await;
```

Comment:
Make sure here we are returning a `SecretResponse` with the error inside of it, since the client always tries to deserialize the payload of a response as a `SecretResponse`